File: ultravox-mobile/src/screens/HomeScreen.js
import React, { useState, useCallback, useEffect, useRef } from 'react';
import {
  SafeAreaView,
  ScrollView,
  View,
  Text,
  TouchableOpacity,
  StyleSheet,
  StatusBar,
  ActivityIndicator,
  Alert
} from 'react-native';
import { startCall, endCall, Role } from '../lib/callFunctions';
import demoConfig from '../lib/demo-config';

const HomeScreen = () => {
  const [isCallActive, setIsCallActive] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [callStatus, setCallStatus] = useState('Inactive');
  const [error, setError] = useState(null);
  const callTimeoutRef = useRef(null);

  // Safety timeout - if no response in 10 seconds, show error and reset
  useEffect(() => {
    if (isLoading) {
      callTimeoutRef.current = setTimeout(() => {
        if (isLoading) {
          setIsLoading(false);
          setError('The conversation is taking too long to start. Please try again.');
        }
      }, 10000);
    }
    
    return () => {
      if (callTimeoutRef.current) {
        clearTimeout(callTimeoutRef.current);
      }
    }
  }, [isCallActive, isLoading]);

  const handleStartChat = async () => {
    try {
      // Clear any previous errors
      setError(null);
      setIsLoading(true);
      setCallStatus('Starting conversation...');

      // Simplify the call configuration
      const callConfig = {
        systemPrompt: demoConfig.callConfig.systemPrompt,
        model: demoConfig.callConfig.model,
        languageHint: demoConfig.callConfig.languageHint,
        voice: demoConfig.callConfig.voice,
        temperature: demoConfig.callConfig.temperature,
        selectedTools: [] // No tools for simple chat
      };

      // Start the call with cleaner approach
      await startCall({
        onStatusChange: handleStatusChange,
        onTranscriptChange: recordTranscripts, // Ignore transcripts
        onDebugMessage: (msg) => console.log('Debug:', msg)
      }, callConfig);
    } catch (error) {
      console.error('Failed to start chat:', error);
      setIsLoading(false);
      setIsCallActive(false);
      setError(`Failed to start conversation: ${error.message}`);
      setCallStatus(`Error: ${error.message}`);
      
      // Show an alert to ensure the user knows something went wrong
      Alert.alert(
        'Error Starting Conversation',
        `There was a problem starting the conversation: ${error.message}`,
        [{ text: 'OK' }]
      );
    }
  };

  const handleEndChat = async () => {
    try {
      setIsLoading(true);
      setCallStatus('Ending conversation...');
      await endCall();
      setIsCallActive(false);
    } catch (error) {
      console.error('Failed to end chat:', error);
    } finally {
      setIsLoading(false);
      setCallStatus('Inactive');
    }
  };

  const recordTranscripts = (transcript) => {
    
  }

  const handleStatusChange = (status, transcript) => {
    // alert("Call status now: " + status.toString());
    console.log('Call status changed:', status);
    setCallStatus(status);
    
    if (status === 'connecting') {
      setIsCallActive(false);
      setIsLoading(true);
    } else if (status === 'idle' || status === 'listening' || status === 'thinking' || status === 'speaking') {
      setIsCallActive(true);
      setIsLoading(false);
    } else if (status.startsWith('disconnect')) {
      console.log('Trascript so far: ', transcript);
      setIsCallActive(false);
      setIsLoading(false);
    } else {
      error.log("Unknown status received: ", status);
      setIsCallActive(false);
      setIsLoading(false);
    }
  };

  const handleEmergencyReset = () => {
    setError(null);
    setIsLoading(false);
    setIsCallActive(false);
    setCallStatus('Inactive');
    
    // Additional cleanup if needed
    try {
      endCall().catch(e => console.warn('Error during emergency reset:', e));
    } catch (e) {
      console.warn('Error during emergency reset:', e);
    }
  };

  return (
    <SafeAreaView style={styles.safeArea}>
      <StatusBar barStyle="light-content" backgroundColor="#000000" />
      <View style={styles.container} id="container">
        <View style={styles.header}>
          <Text style={styles.title}>Voice Assistant</Text>
          <Text style={styles.subtitle}>
            {isCallActive ? "Call is active" : "Ready to start"}
          </Text>
        </View>

        {/* Conversation Area - Using React Native components */}
        <View style={styles.conversationContainer}>
          <View style={styles.conversationContent}>
            <Text style={styles.conversationText}>
              {isCallActive ? 'Call Active' : 'Start'}
            </Text>
          </View>
          
          {isCallActive && (
            <TouchableOpacity 
              onPress={handleEndChat}
              style={styles.endCallButton}
            >
              <Text style={styles.endCallButtonText}>End Call</Text>
            </TouchableOpacity>
          )}
        </View>

        {/* Controls area */}
        <View style={styles.controlsContainer}>
          {!isCallActive && (
            <TouchableOpacity 
              style={[styles.startButton, isLoading && styles.disabledButton]} 
              onPress={handleStartChat}
              disabled={isLoading}
            >
              {isLoading ? (
                <ActivityIndicator size="small" color="white" />
              ) : (
                <Text style={styles.buttonText}>Start Conversation</Text>
              )}
            </TouchableOpacity>
          )}
          
          {error && (
            <TouchableOpacity 
              style={styles.resetButton} 
              onPress={handleEmergencyReset}
            >
              <Text style={styles.buttonText}>Reset</Text>
            </TouchableOpacity>
          )}
        </View>
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  safeArea: {
    flex: 1,
    backgroundColor: '#121212',
  },
  container: {
    flex: 1,
    padding: 16,
    justifyContent: 'space-between',
  },
  header: {
    paddingVertical: 16,
    marginBottom: 8,
  },
  title: {
    fontSize: 24,
    fontWeight: 'bold',
    color: 'white',
    marginBottom: 8,
  },
  subtitle: {
    fontSize: 16,
    color: '#AAA',
  },
  conversationContainer: {
    flex: 1,
    backgroundColor: '#1E1E1E',
    borderRadius: 8,
    overflow: 'hidden',
    alignItems: 'center',
    justifyContent: 'center',
    position: 'relative',
    marginVertical: 20,
  },
  conversationContent: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  conversationText: {
    color: '#999',
    fontSize: 16,
    textAlign: 'center',
  },
  endCallButton: {
    position: 'absolute',
    bottom: 40,
    width: 120,
    height: 36,
    backgroundColor: '#D32F2F',
    borderRadius: 18,
    alignItems: 'center',
    justifyContent: 'center',
  },
  endCallButtonText: {
    color: 'white',
    fontSize: 14,
    fontWeight: '600',
  },
  controlsContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    padding: 16,
    gap: 8,
  },
  startButton: {
    backgroundColor: '#0B57D0',
    borderRadius: 24,
    paddingVertical: 12,
    paddingHorizontal: 24,
    flex: 1,
    alignItems: 'center',
    minHeight: 48,
    justifyContent: 'center',
  },
  disabledButton: {
    backgroundColor: '#0B57D0AA',
  },
  resetButton: {
    backgroundColor: '#E65100',
    borderRadius: 24,
    paddingVertical: 12,
    paddingHorizontal: 24,
    alignItems: 'center',
    justifyContent: 'center',
    minHeight: 48,
    marginLeft: 8,
    width: 100,
  },
  buttonText: {
    color: 'white',
    fontSize: 16,
    fontWeight: '600',
  },
});

export default HomeScreen; 

File: flutter-android/polyapp/lib/screens/home_screen.dart
import 'package:flutter/material.dart';
import '../services/ultravox_service.dart';
import '../models/demo_config.dart';
import '../widgets/mic_toggle_button.dart';

class HomeScreen extends StatefulWidget {
  const HomeScreen({Key? key}) : super(key: key);

  @override
  State<HomeScreen> createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> {
  bool _isCallActive = false;
  bool _isLoading = false;
  String _callStatus = 'Inactive';
  String? _error;

  @override
  void initState() {
    super.initState();
  }

  Future<void> _handleStartChat() async {
    try {
      // Clear any previous errors
      setState(() {
        _error = null;
        _isLoading = true;
        _callStatus = 'Starting conversation...';
      });

      // Get the demo configuration
      final callConfig = DemoConfig.defaultConfig.callConfig;

      // Start the call with status change callback
      await UltravoxService.startCall(
        callConfig,
        _handleStatusChange,
      );
    } catch (error) {
      print('Failed to start chat: $error');
      setState(() {
        _isLoading = false;
        _isCallActive = false;
        _error = 'Failed to start conversation: $error';
        _callStatus = 'Error: $error';
      });

      // Show an alert dialog
      if (mounted) {
        showDialog(
          context: context,
          builder: (context) => AlertDialog(
            title: const Text('Error Starting Conversation'),
            content: Text('There was a problem starting the conversation: $error'),
            actions: [
              TextButton(
                onPressed: () => Navigator.of(context).pop(),
                child: const Text('OK'),
              ),
            ],
          ),
        );
      }
    }
  }

  Future<void> _handleEndChat() async {
    try {
      setState(() {
        _isLoading = true;
        _callStatus = 'Ending conversation...';
      });
      await UltravoxService.endCall();
      setState(() {
        _isCallActive = false;
        _isLoading = false;
        _callStatus = 'Inactive';
      });
    } catch (error) {
      print('Failed to end chat: $error');
      setState(() {
        _isLoading = false;
      });
    }
  }

  void _handleStatusChange(UltravoxService.CallStatus status) {
    print('Call status changed: $status');
    
    setState(() {
      switch (status) {
        case UltravoxService.CallStatus.connecting:
          _isCallActive = false;
          _isLoading = true;
          _callStatus = 'connecting';
          break;
        case UltravoxService.CallStatus.listening:
        case UltravoxService.CallStatus.thinking:
        case UltravoxService.CallStatus.speaking:
          _isCallActive = true;
          _isLoading = false;
          _callStatus = status.toString().split('.').last;
          break;
        case UltravoxService.CallStatus.disconnected:
          _isCallActive = false;
          _isLoading = false;
          _callStatus = 'disconnected';
          break;
        case UltravoxService.CallStatus.error:
          _isCallActive = false;
          _isLoading = false;
          _callStatus = 'error';
          break;
        default:
          _isCallActive = false;
          _isLoading = false;
          _callStatus = 'unknown';
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: const Color(0xFF121212),
      body: SafeArea(
        child: Padding(
          padding: const EdgeInsets.all(16.0),
          child: Column(
            children: [
              // Header
              Container(
                padding: const EdgeInsets.symmetric(vertical: 16.0),
                margin: const EdgeInsets.only(bottom: 8.0),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    const Text(
                      'Voice Assistant',
                      style: TextStyle(
                        fontSize: 24.0,
                        fontWeight: FontWeight.bold,
                        color: Colors.white,
                      ),
                    ),
                    const SizedBox(height: 8.0),
                    Text(
                      _isCallActive ? 'Call is active' : 'Ready to start',
                      style: TextStyle(
                        fontSize: 16.0,
                        color: Colors.grey[400],
                      ),
                    ),
                  ],
                ),
              ),

              // Conversation Area
              Expanded(
                child: Container(
                  decoration: BoxDecoration(
                    color: const Color(0xFF1E1E1E),
                    borderRadius: BorderRadius.circular(8.0),
                  ),
                  width: double.infinity,
                  margin: const EdgeInsets.symmetric(vertical: 20.0),
                  child: Stack(
                    alignment: Alignment.center,
                    children: [
                      // Conversation content
                      Center(
                        child: Text(
                          _isCallActive ? 'Call Active' : 'Start',
                          style: TextStyle(
                            color: Colors.grey[600],
                            fontSize: 16.0,
                          ),
                        ),
                      ),
                      
                      // End call button (when active)
                      if (_isCallActive)
                        Positioned(
                          bottom: 40.0,
                          child: InkWell(
                            onTap: _handleEndChat,
                            child: Container(
                              width: 120.0,
                              height: 36.0,
                              decoration: BoxDecoration(
                                color: const Color(0xFFD32F2F),
                                borderRadius: BorderRadius.circular(18.0),
                              ),
                              alignment: Alignment.center,
                              child: const Text(
                                'End Call',
                                style: TextStyle(
                                  color: Colors.white,
                                  fontSize: 14.0,
                                  fontWeight: FontWeight.w600,
                                ),
                              ),
                            ),
                          ),
                        ),
                    ],
                  ),
                ),
              ),

              // Controls area
              Container(
                padding: const EdgeInsets.all(16.0),
                child: Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    if (!_isCallActive)
                      Expanded(
                        child: InkWell(
                          onTap: _isLoading ? null : _handleStartChat,
                          child: Container(
                            height: 48.0,
                            decoration: BoxDecoration(
                              color: _isLoading
                                  ? const Color(0xFF0B57D0).withOpacity(0.7)
                                  : const Color(0xFF0B57D0),
                              borderRadius: BorderRadius.circular(24.0),
                            ),
                            alignment: Alignment.center,
                            child: _isLoading
                                ? const SizedBox(
                                    width: 20.0,
                                    height: 20.0,
                                    child: CircularProgressIndicator(
                                      strokeWidth: 2.0,
                                      valueColor:
                                          AlwaysStoppedAnimation<Color>(Colors.white),
                                    ),
                                  )
                                : const Text(
                                    'Start Conversation',
                                    style: TextStyle(
                                      color: Colors.white,
                                      fontSize: 16.0,
                                      fontWeight: FontWeight.w500,
                                    ),
                                  ),
                          ),
                        ),
                      ),
                    if (_isCallActive)
                      MicToggleButton(isCallActive: _isCallActive),
                  ],
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
} 